ICE when using a constant as a pattern

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

Bisected to 8c392966a013fd8a09e6b78b3c8d6e442bc278e1 (#137848).

### Code

```Rust
struct Foo<const V: u8>;

trait Bar<'b> {
    fn bar(n: u8);
}

// ICE *STILL* triggered if Foo doesn't have a const param
// ICE *NOT* if Bar doesn't have a lifetime param
// ICE *NOT* triggered obviously with correct `v if v == V`
impl<'b, const V: u8> Bar<'b> for Foo<V> {
    fn bar(n: u8) {
        match n {
            V => (),
            _ => (),
        }
    }
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.87.0-nightly (f9e0239a7 2025-03-04)
binary: rustc
commit-hash: f9e0239a7bc813b4aceffc7f069f4797cde3175c
commit-date: 2025-03-04
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0
```

### Error output

```
error[E0158]: constant parameters cannot be referenced in patterns
  --> src/lib.rs:13:13
   |
10 | impl<'b, const V: u8> Bar<'b> for Foo<V> {
   |          ----------- constant defined here
...
13 |             V => (),
   |             ^ can't be used in patterns


```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at compiler/rustc_mir_build/src/thir/util.rs:25:14:
Box<dyn Any>
stack backtrace:
   0:     0x7fc2987e4da5 - std::backtrace::Backtrace::create::hfef1a8f8aac33464
   1:     0x7fc296bc9ff5 - std::backtrace::Backtrace::force_capture::h8687f78d397650db
   2:     0x7fc295d30dee - std[2cce8494628f4585]::panicking::update_hook::<alloc[ea6342a0a5a8b9b1]::boxed::Box<rustc_driver_impl[e161daf9ab2d11d7]::install_ice_hook::{closure#1}>>::{closure#0}
   3:     0x7fc296be2903 - std::panicking::rust_panic_with_hook::h2c8894f8997fcf6a
   4:     0x7fc295d6d471 - std[2cce8494628f4585]::panicking::begin_panic::<rustc_errors[e90c28bb776fdfc0]::ExplicitBug>::{closure#0}
   5:     0x7fc295d61886 - std[2cce8494628f4585]::sys::backtrace::__rust_end_short_backtrace::<std[2cce8494628f4585]::panicking::begin_panic<rustc_errors[e90c28bb776fdfc0]::ExplicitBug>::{closure#0}, !>
   6:     0x7fc295d5e209 - std[2cce8494628f4585]::panicking::begin_panic::<rustc_errors[e90c28bb776fdfc0]::ExplicitBug>
   7:     0x7fc295d76e61 - <rustc_errors[e90c28bb776fdfc0]::diagnostic::BugAbort as rustc_errors[e90c28bb776fdfc0]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
   8:     0x7fc296352236 - rustc_middle[2d5b20c791c56ad0]::util::bug::opt_span_bug_fmt::<rustc_span[6699764a695b2e79]::span_encoding::Span>::{closure#0}
   9:     0x7fc296339b5a - rustc_middle[2d5b20c791c56ad0]::ty::context::tls::with_opt::<rustc_middle[2d5b20c791c56ad0]::util::bug::opt_span_bug_fmt<rustc_span[6699764a695b2e79]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  10:     0x7fc2963399cb - rustc_middle[2d5b20c791c56ad0]::ty::context::tls::with_context_opt::<rustc_middle[2d5b20c791c56ad0]::ty::context::tls::with_opt<rustc_middle[2d5b20c791c56ad0]::util::bug::opt_span_bug_fmt<rustc_span[6699764a695b2e79]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  11:     0x7fc2946fb380 - rustc_middle[2d5b20c791c56ad0]::util::bug::bug_fmt
  12:     0x7fc298d70e37 - rustc_mir_build[e0917d4cd90e2920]::thir::util::user_args_applied_to_ty_of_hir_id.cold
  13:     0x7fc297d1d777 - <rustc_mir_build[e0917d4cd90e2920]::thir::pattern::PatCtxt>::lower_variant_or_leaf
  14:     0x7fc297d1dfb2 - <rustc_mir_build[e0917d4cd90e2920]::thir::pattern::PatCtxt>::lower_pat_expr
  15:     0x7fc297d1a272 - <rustc_mir_build[e0917d4cd90e2920]::thir::pattern::PatCtxt>::lower_pattern
  16:     0x7fc297d18e8a - <rustc_mir_build[e0917d4cd90e2920]::thir::cx::ThirBuildCx>::pattern_from_hir
  17:     0x7fc297d12a06 - <rustc_mir_build[e0917d4cd90e2920]::thir::cx::ThirBuildCx>::mirror_expr
  18:     0x7fc297d12801 - <rustc_mir_build[e0917d4cd90e2920]::thir::cx::ThirBuildCx>::mirror_expr
  19:     0x7fc297d2714b - rustc_mir_build[e0917d4cd90e2920]::thir::cx::thir_body
  20:     0x7fc297d26c28 - rustc_query_impl[b4e30529d2c77c57]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[b4e30529d2c77c57]::query_impl::thir_body::dynamic_query::{closure#2}::{closure#0}, rustc_middle[2d5b20c791c56ad0]::query::erase::Erased<[u8; 16usize]>>
  21:     0x7fc297f1d406 - rustc_query_system[cb43291f2de4de55]::query::plumbing::try_execute_query::<rustc_query_impl[b4e30529d2c77c57]::DynamicConfig<rustc_data_structures[38628b744a74ab57]::vec_cache::VecCache<rustc_span[6699764a695b2e79]::def_id::LocalDefId, rustc_middle[2d5b20c791c56ad0]::query::erase::Erased<[u8; 16usize]>, rustc_query_system[cb43291f2de4de55]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[b4e30529d2c77c57]::plumbing::QueryCtxt, true>
  22:     0x7fc2985d96a2 - rustc_query_impl[b4e30529d2c77c57]::plumbing::force_from_dep_node::<rustc_query_impl[b4e30529d2c77c57]::DynamicConfig<rustc_data_structures[38628b744a74ab57]::vec_cache::VecCache<rustc_span[6699764a695b2e79]::def_id::LocalDefId, rustc_middle[2d5b20c791c56ad0]::query::erase::Erased<[u8; 16usize]>, rustc_query_system[cb43291f2de4de55]::dep_graph::graph::DepNodeIndex>, false, false, false>>
  23:     0x7fc2985d956d - <rustc_query_impl[b4e30529d2c77c57]::plumbing::query_callback<rustc_query_impl[b4e30529d2c77c57]::query_impl::thir_body::QueryType>::{closure#0} as core[e3f8cccd6bff4ec3]::ops::function::FnOnce<(rustc_middle[2d5b20c791c56ad0]::ty::context::TyCtxt, rustc_query_system[cb43291f2de4de55]::dep_graph::dep_node::DepNode)>>::call_once
  24:     0x7fc2974259ba - <rustc_query_system[cb43291f2de4de55]::dep_graph::graph::DepGraphData<rustc_middle[2d5b20c791c56ad0]::dep_graph::DepsType>>::try_mark_previous_green::<rustc_query_impl[b4e30529d2c77c57]::plumbing::QueryCtxt>
  25:     0x7fc29789c7fd - rustc_query_system[cb43291f2de4de55]::query::plumbing::ensure_must_run::<rustc_query_impl[b4e30529d2c77c57]::DynamicConfig<rustc_data_structures[38628b744a74ab57]::vec_cache::VecCache<rustc_span[6699764a695b2e79]::def_id::LocalDefId, rustc_middle[2d5b20c791c56ad0]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[cb43291f2de4de55]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[b4e30529d2c77c57]::plumbing::QueryCtxt>
  26:     0x7fc297899a13 - rustc_query_impl[b4e30529d2c77c57]::query_impl::check_unsafety::get_query_incr::__rust_end_short_backtrace
  27:     0x7fc297ad5ac8 - rustc_interface[5fc383e867bdd00e]::passes::run_required_analyses
  28:     0x7fc298332dba - rustc_interface[5fc383e867bdd00e]::passes::analysis
  29:     0x7fc298332d99 - rustc_query_impl[b4e30529d2c77c57]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[b4e30529d2c77c57]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[2d5b20c791c56ad0]::query::erase::Erased<[u8; 0usize]>>
  30:     0x7fc298697faa - rustc_query_system[cb43291f2de4de55]::query::plumbing::try_execute_query::<rustc_query_impl[b4e30529d2c77c57]::DynamicConfig<rustc_query_system[cb43291f2de4de55]::query::caches::SingleCache<rustc_middle[2d5b20c791c56ad0]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[b4e30529d2c77c57]::plumbing::QueryCtxt, true>
  31:     0x7fc2986979a2 - rustc_query_impl[b4e30529d2c77c57]::query_impl::analysis::get_query_incr::__rust_end_short_backtrace
  32:     0x7fc29849fc7d - rustc_interface[5fc383e867bdd00e]::passes::create_and_enter_global_ctxt::<core[e3f8cccd6bff4ec3]::option::Option<rustc_interface[5fc383e867bdd00e]::queries::Linker>, rustc_driver_impl[e161daf9ab2d11d7]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  33:     0x7fc2984e66e0 - rustc_interface[5fc383e867bdd00e]::interface::run_compiler::<(), rustc_driver_impl[e161daf9ab2d11d7]::run_compiler::{closure#0}>::{closure#1}
  34:     0x7fc298337588 - std[2cce8494628f4585]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[5fc383e867bdd00e]::util::run_in_thread_with_globals<rustc_interface[5fc383e867bdd00e]::util::run_in_thread_pool_with_globals<rustc_interface[5fc383e867bdd00e]::interface::run_compiler<(), rustc_driver_impl[e161daf9ab2d11d7]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  35:     0x7fc298337e74 - <<std[2cce8494628f4585]::thread::Builder>::spawn_unchecked_<rustc_interface[5fc383e867bdd00e]::util::run_in_thread_with_globals<rustc_interface[5fc383e867bdd00e]::util::run_in_thread_pool_with_globals<rustc_interface[5fc383e867bdd00e]::interface::run_compiler<(), rustc_driver_impl[e161daf9ab2d11d7]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[e3f8cccd6bff4ec3]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x7fc298339277 - std::sys::pal::unix::thread::Thread::new::thread_start::h6a5943b86725331d
  37:     0x7fc2924a370a - <unknown>
  38:     0x7fc292527aac - <unknown>
  39:                0x0 - <unknown>


rustc version: 1.87.0-nightly (f9e0239a7 2025-03-04)
platform: x86_64-unknown-linux-gnu

query stack during panic:
#0 [thir_body] building THIR for `<impl at src/lib.rs:10:1: 10:41>::bar`
#1 [analysis] running analysis passes on this crate
end of query stack
```

</p>
</details>

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs
@@ -220,6 +220,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     ) {
         debug!("fcx {}", self.tag());
 
+        // Don't write user type annotations for const param types, since we give them
+        // identity args just so that we can trivially substitute their `EarlyBinder`.
+        // We enforce that they match their type in MIR later on.
+        if matches!(self.tcx.def_kind(def_id), DefKind::ConstParam) {
+            return;
+        }
+
         if Self::can_contain_user_lifetime_bounds((args, user_self_ty)) {
             let canonicalized = self.canonicalize_user_type_annotation(ty::UserType::new(
                 ty::UserTypeKind::TypeOf(def_id, UserArgs { args, user_self_ty }),

--- a/compiler/rustc_mir_build/src/thir/cx/mod.rs
+++ b/compiler/rustc_mir_build/src/thir/cx/mod.rs
@@ -207,7 +207,7 @@ impl<'tcx> ThirBuildCx<'tcx> {
         &self,
         hir_id: HirId,
     ) -> Option<ty::CanonicalUserType<'tcx>> {
-        crate::thir::util::user_args_applied_to_ty_of_hir_id(self.typeck_results, hir_id)
+        crate::thir::util::user_args_applied_to_ty_of_hir_id(self.tcx, self.typeck_results, hir_id)
     }
 }
 

--- a/compiler/rustc_mir_build/src/thir/pattern/mod.rs
+++ b/compiler/rustc_mir_build/src/thir/pattern/mod.rs
@@ -539,7 +539,7 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
         &self,
         hir_id: hir::HirId,
     ) -> Option<ty::CanonicalUserType<'tcx>> {
-        crate::thir::util::user_args_applied_to_ty_of_hir_id(self.typeck_results, hir_id)
+        crate::thir::util::user_args_applied_to_ty_of_hir_id(self.tcx, self.typeck_results, hir_id)
     }
 
     /// Takes a HIR Path. If the path is a constant, evaluates it and feeds

--- a/compiler/rustc_mir_build/src/thir/util.rs
+++ b/compiler/rustc_mir_build/src/thir/util.rs
@@ -1,12 +1,16 @@
+use std::assert_matches::assert_matches;
+
 use rustc_hir as hir;
+use rustc_hir::def::DefKind;
 use rustc_middle::bug;
-use rustc_middle::ty::{self, CanonicalUserType};
+use rustc_middle::ty::{self, CanonicalUserType, TyCtxt};
 use tracing::debug;
 
 /// Looks up the type associated with this hir-id and applies the
 /// user-given generic parameters; the hir-id must map to a suitable
 /// type.
 pub(crate) fn user_args_applied_to_ty_of_hir_id<'tcx>(
+    tcx: TyCtxt<'tcx>,
     typeck_results: &ty::TypeckResults<'tcx>,
     hir_id: hir::HirId,
 ) -> Option<CanonicalUserType<'tcx>> {
@@ -16,7 +20,23 @@ pub(crate) fn user_args_applied_to_ty_of_hir_id<'tcx>(
     let ty = typeck_results.node_type(hir_id);
     match ty.kind() {
         ty::Adt(adt_def, ..) => {
+            // This "fixes" user type annotations for tupled ctor patterns for ADTs.
+            // That's because `type_of(ctor_did)` returns a FnDef, but we actually
+            // want to be annotating the type of the ADT itself. It's a bit goofy,
+            // but it's easier to adjust this here rather than in the path lowering
+            // code for patterns in HIR.
             if let ty::UserTypeKind::TypeOf(did, _) = &mut user_ty.value.kind {
+                // This is either already set up correctly (struct, union, enum, or variant),
+                // or needs adjusting (ctor). Make sure we don't start adjusting other
+                // user annotations like consts or fn calls.
+                assert_matches!(
+                    tcx.def_kind(*did),
+                    DefKind::Ctor(..)
+                        | DefKind::Struct
+                        | DefKind::Enum
+                        | DefKind::Union
+                        | DefKind::Variant
+                );
                 *did = adt_def.did();
             }
             Some(user_ty)

--- a/tests/crashes/138048.rs
+++ b/tests/crashes/138048.rs
@@ -1,8 +1,0 @@
-//@ known-bug: #138048
-struct Foo;
-
-impl<'b> Foo {
-    fn bar<const V: u8>() {
-        let V;
-    }
-}

--- a/tests/ui/const-generics/generic_const_parameter_types/bad-param-in-pat.rs
+++ b/tests/ui/const-generics/generic_const_parameter_types/bad-param-in-pat.rs
@@ -1,0 +1,10 @@
+struct Foo<'a>(&'a ());
+
+// We need a lifetime in scope or else we do not write a user type annotation as a fast-path.
+impl<'a> Foo<'a> {
+    fn bar<const V: u8>() {
+        let V;
+        //~^ ERROR constant parameters cannot be referenced in patterns
+    }
+}
+fn main() {}

--- a/tests/ui/const-generics/generic_const_parameter_types/bad-param-in-pat.stderr
+++ b/tests/ui/const-generics/generic_const_parameter_types/bad-param-in-pat.stderr
@@ -1,0 +1,11 @@
+error[E0158]: constant parameters cannot be referenced in patterns
+  --> $DIR/bad-param-in-pat.rs:6:13
+   |
+LL |     fn bar<const V: u8>() {
+   |            ----------- constant defined here
+LL |         let V;
+   |             ^ can't be used in patterns
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0158`.

